Adding pagination to Community page

emcoding opened the pull request:
I added paginating to the community (= all users) page. 
I chose not to add a if #users.any? , because there always ARE users, i.e. the organizers. Objections anyone?

@roosdebildt : Could you have a look, please: how many users do we want on one page?  
(If you wait for Pull Request #6 , you have data to play with.) At the Activity page, it has been set to 50, which I think is too many.  
Plus: I copied a bootstrap 'paginator' into the slim file, the same as used elsewhere on the site. 
That means the bootstrap paginator formatting info is in 3 different places. Maybe we can put them together in one place.

emcoding committed:
Adding pagination to Community page

carpodaster commented on app/models/user.rb:
```diff
@@ -40,6 +40,8 @@ class User < ActiveRecord::Base
 
   devise :omniauthable
 
+  paginates_per 10 #MAUD
```
I would not add the pagination configuration here because it adds an (in my opinion) unneeded coupling between the persistence and the presentation layer. The model shouldn't care how clients to it present it. I know kaminari supports it but I think it's a bad design decision on their side. I think there is a `per` scope you can call on the `page` scope (as called in the controller) but that's just serving from memory.

emcoding replied:
Aha, I see what you mean about cluttering the model, thanks.

emcoding committed:
Moving paging config out of model into initializer

emcoding commented:
coaches: I did not use a scope (especially because I couldn't found how to). Instead, I added a kaminari_config.rb where I set a default_per_page. At least, the user.rb model is clean. 

Possible gotcha: In two other models (Activities, Mailer) the default is overridden, so I better find out how to set a scope in their controllers. 

In users_controller I added `.page(params[:page])` like this

``` ruby
 @users = User.ordered(params[:sort], params[:direction]).group('users.id').with_all_associations_joined.page(params[:page])
```

Kaminari has a `per` scope like this: `.per(50)` . 

What to do with it?

yorickpeterse commented:
If I remember correctly you can define a scope as following:

```
class User < ActiveRecord::Base
  scope :paginated, proc { per(50) }
end
```

emcoding commented:
@YorickPeterse Thanks! That helps a lot.

emcoding committed:
Adding page.per()-scope to users controller

emcoding commented:
The neat & simple scoping with `.per(16)` works now. Problem was that the very long line got corrupted by my editor.

yorickpeterse commented on app/controllers/users_controller.rb:
```diff
@@ -12,7 +12,9 @@ def index
     }
-    @users = User.ordered(params[:sort], params[:direction]).group('users.id').with_all_associations_joined
+    @users = User.ordered(params[:sort],
+                          params[:direction]).group('users.id').with_all_associations_joined.
```
Putting a dot at the end of the line is super confusing as it's easy to miss, something along these lines is more readable:

``` ruby
@users = User.ordered(params[:sort], params[:direction])
  .group('users.id')
  .with_all_assocations_joined
  .page(params[:page])
  .per(16)
```

It takes a few more lines but line count in itself isn't really an issue.

emcoding replied:
That's way better! Thanks. 
I take it you do agree with the scope added here inline?

yorickpeterse replied:
Not sure, what's the reason for not using a scope here? Also, out of curiosity why is it limiting to 16 rows per page? It's a bit of an odd number (usually people stick with some multiple of 10).

emcoding replied:
This is the way I understood it from the gem's instructions. They called the `.per()` a scope as well. And they put it chained to the `ordered` method. 
I figured this was the safer way to go. 
16 is mostly arbitrary; my front end guru will decide on a default value for the whole site.:-)

emcoding committed:
Better formatting

yorickpeterse commented:
As it is now this PR looks good to me :+1:

emcoding commented:
Thnx :-)

Here, I removed the pagination configuration out of the model and into the controller. 
However, there are two other controllers (Mailer and Activities) using pagination, where the configuration is still in the model. 
I could set a default value to the Kanimari.config. And then remove the configurations from all the models.

Shall I? ( @carpodaster  ?)

carpodaster commented:
> I could set a default value to the Kanimari.config. And then remove the configurations from all the models. 

Sounds like a good thing, yes! But let's not do this in this PR to focus on one feature/improvement at a time.

emcoding commented:
Ouch! I ran the tests, and it FAILS.
The method that fails is `renders a list of users` in `/spec/views/users/index.html.slim_spec.rb`

``` ruby
require 'spec_helper'

describe 'users/index' do
  before(:each) do
    assign(:filters, { all: 'All', foo: 'Bar' })
    assign(:users, [
      stub_model(User, name: 'Name', email: 'Email', location: 'Location', bio: 'Bio', homepage: 'Homepage', role: 'coach'),
      stub_model(User, name: 'Name', email: 'Email', location: 'Location', bio: 'Bio', homepage: 'Homepage', role: 'coach')
    ])
    allow(view).to receive(:current_season).and_return(build_stubbed(:season))
    allow(controller).to receive(:current_user).and_return(double.as_null_object)
  end

  it 'renders a list of users' do
    render
    assert_select 'tr>td', text: 'Name', count: 2
  end
end
```

This is the test result:

``` ruby

ActionView::Template::Error: undefined method `entry_name' for #<Array:0x007fa7327b9130>
/Users/maud/.rvm/gems/ruby-2.2.2/gems/kaminari-0.16.3/lib/kaminari/helpers/action_view_extension.rb:91:in `page_entries_info'
./app/views/users/index.html.slim:43:in `_app_views_users_index_html_slim__2984173420978787452_70178083746020'
/Users/maud/.rvm/gems/ruby-2.2.2/gems/actionview-4.2.2/lib/action_view/template.rb:145:in `block in render'
/Users/maud/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.2/lib/active_support/notifications.rb:164:in `block in instrument'
/Users/maud/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.2/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/Users/
```

What do I do about it?

The Kaminari helper method (If that is the right word) `page_entries_info` is called in the `app/views/users/index.html.slim`

``` ruby
....
p.pagination-info
  = page_entries_info @users
  = paginate @users, theme: 'twitter-bootstrap-3', pagination_class: 'pagination-sm'
```

 from `gems/kaminari-0.16.3/lib/kaminari/helpers/action_view_extension.rb`

Is it the test, or is it something to do with the gem?

andremedeiros commented:
I think it has to do with the fact that the pagination expects to work on an `ActiveRecord::Relation` whereas the specs create an `Array`, but I might be wrong.

emcoding commented:
EDIT No wait, I think I found something.  

 Ouch ouch. I don't understand what to do. Other than writing a new test.

emcoding commented:
UPDATE 
1 There is a gem available which help solving the RSpecs. https://github.com/nessche/kaminari-rspec 
2 In the other views which use the Kaminari pagination, there are no spec tests on the view
3 Maybe cursing in the test-church, but I don't see much use for this test. What am I missing?
(It creates 2 users, and test if they are listed.)

So. What should I do? Any advice?

yorickpeterse commented:
Personally I never bother testing views because they are far too fragile. Simply moving `div`s around or renaming classes can break a whole bunch of tests when that doesn't strictly mean something is actually broken.

Looking at this spec itself it's not terribly useful, so either it would have to be extended to test more than just the amount of `tr>td` occurrences or it should be removed.

carpodaster commented:
I'm with @YorickPeterse. I put `render_views` in the controller tests to catch syntax errors in views but that's as far as it gets in terms of view testing for me.

emcoding commented:
So, can I delete the test? And put the controller test in my backlog ;-) for when I am able to write tests that matter.

yorickpeterse commented:
@F3PiX Yes, that should be fine.

emcoding commented:
Trallalieetralala.

emcoding committed:
Deleting incomplete view-spec test

yorickpeterse commented:
Looks good :shipit:
